kedro/io/data_catalog.py

```diff
@@ -615,15 +616,55 @@ def add_transformer(
                 )
             self._transformers[data_set_name].append(transformer)
 
-    def list(self) -> List[str]:
-        """List of ``DataSet`` names registered in the catalog.
+    def list(
+        self, regex_search: Optional[str] = None, get_objects: bool = False
+    ) -> Union[List[str], Dict[str, AbstractDataSet]]:
+        """
+        List of ``DataSet`` names registered in the catalog if called with
+        default arguments. Additional facility to limit dataset returned by
+        a regular expression name as well as retrieving a dictionary of dataset
+        name / object pairs.
+
+        Args:
+            regex_search: An optional regular expression which can be provided
+                to limit the data sets returned by a particular pattern
+            get_objects: An optional flag to rerun a dictionary with the actual
+            ``DataSet`` objects in addition to the names as keys
 
         Returns:
-            A List of ``DataSet`` names, corresponding to the entries that are
-            registered in the current catalog object.
+            A list of ``DataSet`` names or a dictionary of name/object pairs
+            depending on the arguments provided
+
+        Raises:
+            SyntaxError: When invalid regex filter is provided
+
+        Example:
+        ::
 
+            >>> io = DataCatalog()
+            >>> # get data sets where the substring 'raw' is present
+            >>> raw_data = io.list(regex_search='raw')
+            >>> # get data sets which start with 'prm' or 'feat'
+            >>> feat_eng_data = io.list(regex_search='^(prm|feat)')
         """
-        return list(self._data_sets.keys())
+
+        if regex_search is not None:
+            try:
+                pattern = re.compile(regex_search, flags=re.IGNORECASE)
+            except re.error:
+                raise SyntaxError(
+                    f"Invalid pattern regular expression provided: '{regex_search}'"
```

Comment:
```suggestion
                    f"Invalid pattern regular expression provided: `{regex_search}`"
```

I believe that as a rule, we try and use backticks in error messages, so to be consistent^